### DISCUSSION TITLE
Fastapi app.mount for static files not working behind proxy

### DISCUSSION THREAD

**Discussion by vg249** [fastapi/fastapi#12151] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI(
    root_path="/api"
)

app.mount(
    "/reference",
    staticfiles.StaticFiles(directory="./api_reference/site/", html=True),
    name="reference",
)
```

### Description

I am mounting my static files like below and have `root_path` defined in app declaration as `/api`,

```
app.mount(
   "/reference",
   staticfiles.StaticFiles(directory="./api_reference/site/", html=True),
   name="reference"
)
```
Application is running behind a proxy. 

`/api/reference` is not working and gets redirected to `/reference`.

### Operating System

Linux

### FastAPI Version

0.113.0

### Pydantic Version

2.8.2

### Python Version

3.10.14

### Additional Context

This is already noted in https://github.com/fastapi/fastapi/discussions/11033#discussioncomment-8546202, but no appropriate solution is provided.

**Comment by akash-manoharan** [comment]
Is there any update for this issue? We're also facing the similar issue in linux machine.

**Comment by YuriiMotov** [comment]
Is it possible to configure your proxy not to strip prefix? So that the path passed to app would start with `/api` prefix

**Comment by martinberoiz** [comment]
I found this same problem, I was surprised that this doesn't work as expected. I will have to resort to creating a route:

```py
@router.get("/reference/{file_path:path}", include_in_schema=False)
async def static_file(file_path: str):
    ...
```

**Comment by gitsult4n** [comment]
Known gap: setting `root_path="/api"` as a static app param only affects OpenAPI/docs URL generation, not the actual redirect behavior generated inside a mounted Starlette `StaticFiles` sub-app — its trailing-slash redirect builds the path from the raw scope, and in some Starlette versions that dropped the prefix instead of respecting it.

Fix per the official FastAPI "Behind a Proxy" docs: don't rely on the static `root_path=` param alone — use an ASGI middleware that reads the `X-Forwarded-Prefix` header from your proxy and sets `scope["root_path"]` per request dynamically before the app runs. Your proxy needs to actually send that header, and if you're on uvicorn you'll also need `--proxy-headers`.

Also worth checking your Starlette version — the "trailing-slash redirect drops the prefix" behavior has had fixes in later releases, so upgrading FastAPI/Starlette first is worth trying.

**Reply by YuriiMotov** [reply]
This is not fully correct - `root_path="/api"` just implies that your proxy shouldn't strip the path prefix. Then everything works. The issue is in FastAPI docs that state that `root_path` parameter is equal to `--root-path` option of Uvicorn. They are not equal.

**Reply by gitsult4n** [reply]
You are right and my earlier comment was wrong for current versions

Measured on fastapi 0.141.1 starlette 1.6.0 uvicorn 0.52.3 with a `/{p:path}` route echoing the scope

| setup | scope root_path | scope path when the client asks for `/thing` |
|---|---|---|
| `FastAPI(root_path="/api")` | `/api` | `/thing` untouched |
| `uvicorn --root-path /api` | `/api` | `/api/thing` prefix put back |

The app param wants the proxy to keep the prefix
The uvicorn flag assumes the proxy already stripped it and adds it back to satisfy the asgi spec
Send a prefixed path into `--root-path /api` and the scope path becomes `/api/api/thing` and you get a 404
So the docs line claiming the two are equal is wrong in the way you describe

Why plain routes hide it and mounts expose it
`Mount` sets the child root_path to app root_path plus mount path

```
FastAPI(root_path="/api") + app.mount("/static", ...)

GET /api/static/x.txt -> child_root_path=/api/static  child_path=/api/static/x.txt -> 200
GET /static/x.txt     -> child_root_path=/api/static  child_path=/static/x.txt     -> 404
```

A child strips its own root_path only when the path starts with it
With the prefix stripped `StaticFiles` ends up looking for `static/x.txt` inside the directory
`/ping` answers either way which is why `root_path` looks fine until something is mounted

All four combinations with `app.mount("/static", StaticFiles(directory="static"))`

| proxy | config | `/static/x.txt` |
|---|---|---|
| keeps `/api` | `FastAPI(root_path="/api")` | 200 |
| strips `/api` | `FastAPI(root_path="/api")` | 404 |
| strips `/api` | `uvicorn --root-path /api` | 200 |
| keeps `/api` | `uvicorn --root-path /api` | 404 |

If the proxy strips the prefix and you cannot pass the flag then set both fields yourself

```python
class PrefixMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] == "http":
            prefix = dict(scope["headers"]).get(b"x-forwarded-prefix", b"").decode()
            if prefix:
                scope["root_path"] = prefix
                if not scope["path"].startswith(prefix):
                    scope["path"] = prefix + scope["path"]
        await self.app(scope, receive, send)
```

Setting `root_path` alone is exactly what breaks the mount
Prepending it to `path` as well gives 200 on `/static/x.txt` and `servers: [{"url": "/api"}]` in the schema

**Comment by sunilkumarvalmiki** [comment]
The important point here is to make the proxy-to-ASGI path contract explicit. There are two valid deployment layouts, and the configuration must match the one your proxy actually uses.

**1. The proxy preserves the /api prefix**

The ASGI application receives paths such as /api/reference and the application is configured with the external root path:

```python
app = FastAPI(root_path="/api")
```

Start Uvicorn without adding the same prefix again. In this layout, the proxy should forward the /api prefix unchanged. The mounted StaticFiles application should then receive a scope consistent with that contract.

**2. The proxy strips the /api prefix**

The ASGI application receives /reference instead of /api/reference. In that case, let the server provide the external root path:

```bash
uvicorn main:app --root-path /api
```

For this layout, avoid also configuring FastAPI(root_path="/api") blindly. The FastAPI application parameter and Uvicorn's --root-path option affect the ASGI scope differently; combining them with a prefix-stripping proxy can make the prefix appear twice or leave the mounted application with an inconsistent scope.

This also explains why ordinary API routes may work while a mounted StaticFiles route fails. A mount is a child ASGI application. It combines the mount path with root_path and uses the incoming scope when resolving files and generating the trailing-slash redirect. If /api is removed by the proxy but the application is configured as though it is still present, a request to /api/reference can be redirected to /reference instead of /api/reference/.

I would verify the contract with a temporary diagnostic middleware:

```python
@app.middleware("http")
async def log_scope(request: Request, call_next):
    print({
        "root_path": request.scope.get("root_path"),
        "path": request.scope.get("path"),
        "raw_path": request.scope.get("raw_path"),
        "x_forwarded_prefix": request.headers.get("x-forwarded-prefix"),
    })
    return await call_next(request)
```

Then test all of these separately:

1. /api/reference — inspect the Location header if it redirects.
2. /api/reference/ — confirm the mounted index is served.
3. /api/reference/index.html — confirm a known static file is found.
4. The same paths directly against the upstream service, bypassing the proxy.

If the Location header changes from /api/reference to /reference, the first thing to fix is the prefix contract between the proxy and ASGI scope—not the static directory. Also check that the proxy forwards the prefix consistently, that trusted forwarded-header handling is configured intentionally, and that your exact FastAPI/Starlette/Uvicorn versions are compatible. The official FastAPI proxy guidance is useful for validating the deployment contract: https://fastapi.tiangolo.com/advanced/behind-a-proxy/
